src/server/native_controllers/index.js

```diff
@@ -0,0 +1,117 @@
+import path from 'path';
+import { get } from 'lodash';
+import { Observable } from 'rxjs';
+import Joi from 'joi';
+import { spawnNativeController } from './spawn_native_controller';
+import { findPluginSpecs } from '../../plugin_discovery';
+import { safeChildProcess } from '../../utils/child_process';
+import { createNativeController } from './native_controller';
+import { getNativeControllerConfig } from './native_controller_config';
+
+const specSchema = Joi.object().keys({
+  pluginId: Joi.string().required(),
+  path: Joi.string().required(),
+  config: Joi.array().items(Joi.string())
+});
+
+const getNativeControllers = async (kbnServer) => {
+  const {
+    packageJson$,
+  } = findPluginSpecs(kbnServer.settings);
+
+  const spec$ = packageJson$
+    .mergeMap(packageJson => {
+      const nativeControllerSpecs = get(packageJson.contents, 'kibana.nativeControllers');
+
+      if (!nativeControllerSpecs) {
+        return Observable.empty();
+      }
+
+      return nativeControllerSpecs.map(spec => {
+        const result = Joi.validate(spec, specSchema);
+        if (result.error) {
+          throw result.error;
+        }
+
+        return {
+          pluginId: spec.pluginId,
+          path: path.resolve(packageJson.directoryPath, spec.path),
+          config: spec.config
+        };
+      });
+    });
+
+  const nativeControllerAndConfig$ = spec$
+    .mergeMap(async spec => {
+      const nativeControllerProcess = spawnNativeController(spec.path);
+      return {
+        nativeController: createNativeController(spec.pluginId, nativeControllerProcess),
+        specConfig: spec.config,
+      };
+    });
+
+  return await nativeControllerAndConfig$
+    .do(({ nativeController }) => {
+      const childProcessStatus = safeChildProcess(nativeController.process);
+      nativeController.process.on('exit', (code) => {
+        if (childProcessStatus.terminating || nativeController.killed) {
+          return;
+        }
+
+        // this can happen before we have the logging configured, so we log to both
+        // console.error and server.log to ensure something is output somewhere.
+        const message = `${nativeController.pluginId}'s native controller exited with code ${code}`;
+        console.error(message);
```

Comment:
Under what conditions can this fire before logging is configured?

If we can determine that logging *is* configured, we should conditionally log to console.error based on whether server.log is available rather than always logging to console.error.